src/index.js

```diff
@@ -0,0 +1,234 @@
+    }
+  }
+
+  /**
+    @function push
+    Add a new chain link to the end of the execution chain
+    @param chainLink {Object} - chain link implementation
+    @return {number} - index of the chain link in the execution chain
+  **/
+  push(chainLink) {
+    assert(chainLink);
+    assert(chainLink.handle, 'Chain link must implement handle(message) function');
+    assert(chainLink.next, 'Chain link must implement next() function');
+    assert(chainLink.link, 'Chain link must implement a link(next) function');
+    const index = this.chainLinks.push(chainLink) - 1;
+    this.chainStart = this.chainLinks[0];
+    this.chainEnd = this.chainLinks[this.chainLinks.length - 1];
+    return index;
+  }
+
+  /**
+    @function bindAdapter
+    Add a new adapter to the exposed module instance
+    @param adapterName {string} - label for the adapter property in the module instance
+    @param adapterInstance {Object} - instance of an adapter
+
+    Example:
+      const { chain } = require(..);
+      chain.bindAdapter('unicorn', new MyUnicorn(..));
+
+      const { unicorn, chain } = require(..);
+      unicorn.yay() // unicorn is now defined
+  **/
+  bindAdapter(adapterName, adapterInstance) {
+    assert(adapterName);
+    assert(adapterInstance);
+    if (!this.adapters.has(adapterName) && !instance[adapterName]) {
+      this.adapters.set(adapterName, adapterInstance);
+      instance[adapterName] = adapterInstance;
+    }
+  }
+
+  /**
+    @function unbindAdapter
+    Remove adapter from the exposed module instance
+    @param adapterName
+  **/
+  unbindAdapter(adapterName) {
+    assert(adapterName);
+    if (this.adapters.has(adapterName)) {
+      delete instance[adapterName];
+      this.adapters.delete(adapterName);
+    }
+  }
+}
+
+/**
+ * Configure and return a chain of processing a log message. It should be required on module launch.
+ * It sets up loggers, notifiers for bugs, and exposes the internal providers, abstracted.
+ * If no parameter is provided, it will return already configured providers
+ * @param  {Object}  [config] Configuration object required by the internal providers
+ * @param  {Array}   [config.chainLinks] Array of instances of chainLinks
+ * @param  {Object}  [config.adapters] Object with adapters to be exposed
+ * @param  {boolean} [config.CONSOLE_LOGGING] Switches on/off the console logging chain
+ * @param  {boolean} [config.MIN_LOG_LEVEL] Minimal log level for a message to be processed
+ * @param  {boolean} [config.MIN_LOG_LEVEL_CONSOLE] Minimal log level for a message to be processed by console chain link (more prior over MIN_LOG_LEVEL)
+ * @param  {boolean} [config.DEFAULT_LOG_LEVEL] Fall back if a message does not have one
+ * Note that this module can also be configured with environment variables.
+ * Environment variables have higher priority over a settings object
+ * @return {Object}  The configured providers
+ */
+module.exports = (config) => {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 21 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.dialonce.io/coding_rules#rule_key=javascript%3AS3776)